Don't use Vector<UChar> to build strings

Ahmad-S792 opened the pull request:
#### 6b1eb391bf3d0bf4fa2b3834e61df0a4dfa95c11
<pre>
Don&apos;t use Vector&lt;UChar&gt; to build strings

Don&apos;t use Vector&lt;UChar&gt; to build strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=119319">https://bugs.webkit.org/show_bug.cgi?id=119319</a>

Reviewed by Darin Adler.

Merge - <a href="https://chromium.googlesource.com/chromium/blink/+/9fda929af533c3c1aa3961c4818ea30ed7499021">https://chromium.googlesource.com/chromium/blink/+/9fda929af533c3c1aa3961c4818ea30ed7499021</a>

This patch replaces a bunch of sites which were using Vector&lt;UChar&gt;
to construct string with StringBuffer.

* Source/WebCore/platform/text/LocaleICU.cpp:
(WebCore::LocaleICU::decimalSymbol):
(WebCore::LocaleICU::decimalTextAttribute):
(WebCore::getDateFormatPattern):
(WebCore::LocaleICU::createLabelVector):
(WebCore::getFormatForSkeleton):

Canonical link: <a href="https://commits.webkit.org/258081@main">https://commits.webkit.org/258081@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58243541659f928e64390ffc6d86d7d4ab5d1e9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110149 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170423 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104835 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10928 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/855 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93257 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107994 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8258 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91519 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34875 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22915 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77849 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3686 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24434 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3705 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/814 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9816 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43934 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5542 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5483 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/text/LocaleICU.cpp:
```diff
@@ -88,9 +89,9 @@ String LocaleICU::decimalTextAttribute(UNumberFormatTextAttribute tag)
     if (U_FAILURE(status) && !needsToGrowToProduceBuffer(status))
         return String();
-    Vector<UChar> buffer(bufferLength);
+    StringBuffer<UChar> buffer(static_cast<unsigned>(bufferLength));
```
I believe the `static_cast` here is not needed and should be removed.

darinadler commented on Source/WebCore/platform/text/LocaleICU.cpp:
```diff
@@ -153,9 +154,9 @@ static String getDateFormatPattern(const UDateFormat* dateFormat)
     if (!needsToGrowToProduceBuffer(status) || !length)
         return emptyString();
-    Vector<UChar> buffer(length);
+    StringBuffer<UChar> buffer(static_cast<unsigned>(length));
```
I believe the `static_cast` here is not needed and should be removed.

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/platform/text/LocaleICU.cpp:
```diff
@@ -175,9 +176,9 @@ std::unique_ptr<Vector<String>> LocaleICU::createLabelVector(const UDateFormat*
         if (!needsToGrowToProduceBuffer(status))
             return makeUnique<Vector<String>>();
-        Vector<UChar> buffer(length);
+        StringBuffer<UChar> buffer(static_cast<unsigned>(length));
```
I believe the `static_cast` here is not needed and should be removed.

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/platform/text/LocaleICU.cpp:
```diff
@@ -262,9 +263,9 @@ static String getFormatForSkeleton(const char* locale, const UChar* skeleton, in
     int32_t length = udatpg_getBestPattern(patternGenerator, skeleton, skeletonLength, 0, 0, &status);
     if (needsToGrowToProduceBuffer(status) && length) {
-        Vector<UChar> buffer(length);
+        StringBuffer<UChar> buffer(static_cast<unsigned>(length));
```
I believe the `static_cast` here is not needed and should be removed.

Ahmad-S792 replied:
Done.

darinadler reviewed (approved):
If this eliminates all uses of the `String::adopt` function that takes a vector, then that function should also be removed from the `String` class.

Ahmad-S792 committed:
Don't use Vector<UChar> to build strings

Don't use Vector<UChar> to build strings
https://bugs.webkit.org/show_bug.cgi?id=119319

Reviewed by Darin Adler.

Merge - https://chromium.googlesource.com/chromium/blink/+/9fda929af533c3c1aa3961c4818ea30ed7499021

This patch replaces a bunch of sites which were using Vector<UChar>
to construct string with StringBuffer.

* Source/WebCore/platform/text/LocaleICU.cpp:
(WebCore::LocaleICU::decimalSymbol):
(WebCore::LocaleICU::decimalTextAttribute):
(WebCore::getDateFormatPattern):
(WebCore::LocaleICU::createLabelVector):
(WebCore::getFormatForSkeleton):

Canonical link: https://commits.webkit.org/258081@main
